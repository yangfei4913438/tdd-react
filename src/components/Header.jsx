import React, { Component } from "react";
import "./Header.css";
import { connect } from "react-redux";
import { actions } from "../containers/TodoList/todoStore";

// 映射state
const mapState = (state) => {
  // 使用这种方法，为了方便有需要全局state的操作
  const { todo } = state;
  return {
    value: todo.inputValue,
  };
};
// 映射action
const mapDispatch = (dispatch) => ({
  // 修改输入框内容
  handleInputChange(value) {
    dispatch(actions.changeInputValue(value));
  },
});
/**
 * 封装一下连接方法
 * 注意：这里的 connect 不能被封装成函数返回值，否则运行报错！
 * */
const connStore = connect(mapState, mapDispatch);

/**
 * Header 组件
 * @param value 输入框的默认值
 * */
class Header extends Component {
  handleInputKeyUp = (e) => {
    const { value } = this.props;
    // 13表示回车键
    if (e.keyCode === 13 && value) {
      // 调用回调
      this.props.addUndoItem(value);
      // 重置为空
      this.props.handleInputChange("");
    }
  };

  render() {
    const { value, handleInputChange } = this.props;
    return (
      <div className={"header"}>
        <div className={"header_content"}>
          TodoList
          <input
            className={"header_input"}
            data-test={"header_input"}
            value={value}
            placeholder={"请输入Todo"}
            onChange={(e) => handleInputChange(e.target.value)}
            onKeyUp={this.handleInputKeyUp}
          />
        </div>
      </div>
    );
  }
}

export default connStore(Header);
