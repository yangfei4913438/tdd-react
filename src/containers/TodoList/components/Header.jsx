import React, { Component } from 'react';
import './Header.css'

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ''
    }
  }

  handleInputChange = e => {
    this.setState({
      value: e.target.value
    })
  }

  handleInputKeyUp = (e) => {
    const { value } = this.state;
    if (e.keyCode === 13 && value) {
      // 调用回调
      this.props.addUndoItem(value)
      // 重置为空
      this.setState({
        value: ''
      })
    }
  }

  render() {
    const  { value } = this.state;
    return <div className={'header'}>
      <div className={'header_content'}>
        TodoList
        <input
          className={'header_input'}
          data-test={'input'}
          value={value}
          placeholder={'请输入Todo'}
          onChange={this.handleInputChange}
          onKeyUp={this.handleInputKeyUp}
        />
      </div>

    </div>
  }
}

export default Header
