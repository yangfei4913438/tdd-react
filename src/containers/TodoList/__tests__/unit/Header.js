import React from "react";
import { shallow } from 'enzyme';
import Header from "../../components/Header";

// 定义输入框的DOM变量
let inputElem
// 定义组件对象
let wrapper
// 定义回调方法
let headerCallback
// 单元测试初始化
beforeAll(() => {
  // shallow 是一个浅渲染，只渲染父组件，不会管子组件。一般用于单元测试。
  wrapper = shallow(<Header />)
})
// 每个用例执行之前初始化
beforeEach(() => {
  // 通过属性选择器来定位对象，不要使用类名，避免类名被修改导致的测试用例失效。
  inputElem = wrapper.find('[data-test="input"]');
  // 回调方法重置
  headerCallback = jest.fn();
})

describe('Header 组件', () => {
  test('组件渲染样式不变', () => {
    // 断言快照一致
    // 主要是判断内容是否发生变化，CSS发生变化无法监控
    expect(wrapper).toMatchSnapshot();
  })

  test('组件包含一个 input 框', () => {
    // 断言dom对象存在
    expect(inputElem).toExist();
  })

  test('inout 框, 初始化应该为空', () => {
    // 断言dom存在属性value, 值为空字符串
    // 如果用户不进行任何操作，此时应该是初始状态。
    expect(inputElem).toHaveProp('value', '');
  })

  test('inout 框, 当用户输入时，值会跟随变化', () => {
    const userInput = '今天开始学习jest'
    // 模拟用户输入, 第一个参数是事件类型，第二个参数是event
    inputElem.simulate('change', {
      target: {
        value: userInput
      }
    })
    // 数据测试
    // 断言用户输入的值, 是上面定义的
    expect(wrapper).toHaveState('value', userInput);
    // dom测试
    // dom对象变更后，需要重新取值，否则取出来的就是之前未修改的值
    inputElem = wrapper.find('[data-test="input"]');
    // 断言DOM上的属性，是上面定义的
    expect(inputElem).toHaveProp('value', userInput)
  })

  test('inout 框输入回车时，如果无内容，无操作', () => {
    // 传入了回调方法
    wrapper.setProps({ addUndoItem: headerCallback })
    // 初始化state
    wrapper.setState({ value: '' });
    // dom对象变更后，需要重新取值，否则取出来的就是之前未修改的值
    inputElem = wrapper.find('[data-test="input"]');
    // 触发回车
    inputElem.simulate('keyUp', {
      keyCode: 13
    })
    // 断言，回调方法，没有被调用过
    expect(headerCallback).not.toHaveBeenCalled()
  })

  test('inout 框输入回车时，如果有内容，回调函数应该被调用', () => {
    // 传入了回调方法
    wrapper.setProps({ addUndoItem: headerCallback })
    // dom对象变更后，需要重新取值，否则取出来的就是之前未修改的值
    inputElem = wrapper.find('[data-test="input"]');

    // 定义用户要输入的内容
    const userInput = '今天开始学习jest'
    // 模拟用户输入
    inputElem.simulate('change', {
      target: {
        value: userInput
      }
    })
    // 触发回车
    inputElem.simulate('keyUp', {
      keyCode: 13
    })
    // 断言，回调方法，被调用过，参数是上面模拟的输入
    expect(headerCallback).toHaveBeenCalledWith(userInput);
    // 输入完成之后，输入框应该是重置为空
    // 数据测试
    // 断言用户输入的值, 是上面定义的
    expect(wrapper).toHaveState('value', '');
    // dom测试
    // dom对象变更后，需要重新取值，否则取出来的就是之前未修改的值
    inputElem = wrapper.find('[data-test="input"]');
    // 断言DOM上的属性，是上面定义的
    expect(inputElem).toHaveProp('value', '')
  })
})
