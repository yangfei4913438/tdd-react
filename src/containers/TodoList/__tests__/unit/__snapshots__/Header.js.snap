// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 组件 组件渲染样式不变 1`] = `
<div
  className="header"
>
  <div
    className="header_content"
  >
    TodoList
    <input
      className="header_input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="请输入Todo"
      value=""
    />
  </div>
</div>
`;
